File: cracking_the_coding_interview/a_arrays_and_strings/a/1.1.ts
export {}

// Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

const string_one = "hello";
const string_two = "world";
const string_three = "apple";
const string_four = "12345";
const string_five = "abracadabra";

const isUniqueCharacters = (str: string) => {
  let hash_table = {};
  let isUnique = true;

  str.split("").forEach((letter) => {
    if (Object.keys(hash_table).includes(letter)) {
      isUnique = false;
    } else {
      hash_table[letter] = 1;
    }
  })
  return isUnique;
}

//Optimal Answer
const isUniqueCharactersOptimal = (str: string) => {
  let hash_table = {};
  for (let letter of str) {
    if (hash_table[letter]) {
      return false; // If duplicate character found, return false
    } else {
      hash_table[letter] = true;
    }
  }
  return true; // If no duplicate characters found, return true
};

File: cracking_the_coding_interview/a_arrays_and_strings/a/1.1.py
# Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

string_one = "hello"
string_two = "world"
string_three = "apple"
string_four = "12345"
string_five = "abracadabra"

def isUniqueCharacters(str):
    # Creating a hash table using a dictionary
    hash_table = {}

    for letter in str:
        if not letter in hash_table:
            hash_table[letter] = 1
        else:
            return False
    return True


print(isUniqueCharacters(string_one))
print(isUniqueCharacters(string_two))
print(isUniqueCharacters(string_three))
print(isUniqueCharacters(string_four))
print(isUniqueCharacters(string_five))
